### PR TITLE
fix(ecs): reduce ecs service task role cloudwatch permissions when no log configured (under feature flag)

### PR DESCRIPTION
**Pull request by GavinZZ**
### Issue # (if applicable)

Closes https://github.com/aws/aws-cdk/issues/31397

### Reason for this change

If a FargateService or EC2 Service has `enableExecuteCommand: true` and the ECS cluster it runs on has `executeCommandConfiguration.logging` set to anything but `ecs.ExecuteCommandLogging.NONE` then the CDK automatically grants the underlying TaskDefinition overly broad cloudwatch logs permissions regardless of need even if the logging configuration has no cloudwatch logs config set.

This is not right behaviour as these permissions are not needed. 

### Description of changes

Add feature flag and if feature flag is on and cloudwatch log is configured, reduce the permissions. 

### Description of how you validated changes

New unit tests, integration tests.

### Checklist
- [x] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/aws-cdk-lib/aws-ecs/lib/base/base-service.ts [hunk 1]
```diff
@@ -1033,23 +1033,30 @@ export abstract class BaseService extends Resource
...
-        'logs:CreateLogStream',
-        'logs:DescribeLogStreams',
-        'logs:PutLogEvents',
-      ],
-      resources: [logGroupArn],
-    }));
+    // When Feature Flag is false, keep the previous behaviour for non-breaking changes.
+    // When Feature Flag is true and when cloudwatch log group is specified in logConfiguration, then
+    // append the necessary permissions to the task definition.
+    if (!reducePermissions || logConfiguration?.cloudWatchLogGroup) {
```
Comment (shikha372) [inline packages/aws-cdk-lib/aws-ecs/lib/base/base-service.ts:1042]: If one choose to specify Logging as `OVERRIDE` then providing log configuration is [necessary](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ecs-cluster-executecommandconfiguration.html), can we add a validation to verify this.

File: packages/@aws-cdk-testing/framework-integ/test/aws-ecs-patterns/test/ec2/integ.multiple-application-load-balanced-ecs-service.ts [hunk 2]
```diff
@@ -3,9 +3,14 @@ import { Cluster, ContainerImage } from 'aws-cdk-lib/aws-ecs';
...
 import * as integ from '@aws-cdk/integ-tests-alpha';
 import { ApplicationMultipleTargetGroupsEc2Service } from 'aws-cdk-lib/aws-ecs-patterns';
-import { AUTOSCALING_GENERATE_LAUNCH_TEMPLATE } from 'aws-cdk-lib/cx-api';
+import { AUTOSCALING_GENERATE_LAUNCH_TEMPLATE, REDUCE_EC2_FARGATE_CLOUDWATCH_PERMISSIONS } from 'aws-cdk-lib/cx-api';
 
-const app = new App({ postCliContext: { [AUTOSCALING_GENERATE_LAUNCH_TEMPLATE]: false } });
+const app = new App({
+  postCliContext: {
+    [AUTOSCALING_GENERATE_LAUNCH_TEMPLATE]: false,
+    [REDUCE_EC2_FARGATE_CLOUDWATCH_PERMISSIONS]: false,
```
Comment (shikha372) [inline packages/@aws-cdk-testing/framework-integ/test/aws-ecs-patterns/test/ec2/integ.multiple-application-load-balanced-ecs-service.ts:11]: should we have test for this set to true or without this feature flag, to confirm that we've expected default behaviour in newer versions.

File: packages/@aws-cdk-testing/framework-integ/test/aws-ecs/test/fargate/integ.enable-execute-command.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+taskDefinition.addContainer('web', {
+  image: ecs.ContainerImage.fromRegistry('amazon/amazon-ecs-sample'),
+  healthCheck: {
+    command: ['CMD-SHELL', 'curl localhost:8000'],
+    interval: Duration.seconds(60),
+    timeout: Duration.seconds(40),
+  },
+});
+
+new ecs.FargateService(stack, 'FargateService', {
```
Comment (shikha372) [inline packages/@aws-cdk-testing/framework-integ/test/aws-ecs/test/fargate/integ.enable-execute-command.ts:48]: shall we add for EC2 type as well

File: packages/aws-cdk-lib/aws-ecs/lib/base/base-service.ts (hunk 1 above)
Comment (GavinZZ) [reply packages/aws-cdk-lib/aws-ecs/lib/base/base-service.ts:1042]: Yes, this is already checked from the cluster construct, https://github.com/aws/aws-cdk/blob/main/packages/aws-cdk-lib/aws-ecs/lib/cluster.ts#L228

File: packages/@aws-cdk-testing/framework-integ/test/aws-ecs/test/fargate/integ.enable-execute-command.ts (hunk 3 above)
Comment (GavinZZ) [reply packages/@aws-cdk-testing/framework-integ/test/aws-ecs/test/fargate/integ.enable-execute-command.ts:48]: Yes, good point. I forgot to add an integ test for EC2 service.

File: packages/aws-cdk-lib/aws-ecs/lib/base/base-service.ts [hunk 4]
```diff
@@ -1033,23 +1033,30 @@ export abstract class BaseService extends Resource
   }
 
   private executeCommandLogConfiguration() {
+    const reducePermissions = FeatureFlags.of(this).isEnabled(cxapi.REDUCE_EC2_FARGATE_CLOUDWATCH_PERMISSIONS);
```
Comment (shikha372) [inline packages/aws-cdk-lib/aws-ecs/lib/base/base-service.ts:1036]: can we define the behaviour with or without feature flag in [README here.](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ecs-cluster-executecommandconfiguration.html)

File: packages/aws-cdk-lib/cx-api/lib/features.ts [hunk 5]
```diff
@@ -1126,6 +1127,21 @@ export const FLAGS: Record<string, FlagInfo> = {
     recommendedValue: true,
     compatibilityWithOldBehaviorMd: 'Disable the feature flag to use input and output path fields for s3 URI',
   },
+
+  //////////////////////////////////////////////////////////////////////
+  [REDUCE_EC2_FARGATE_CLOUDWATCH_PERMISSIONS]: {
+    type: FlagType.BugFix,
+    summary: 'When enabled, we will only grant the necessary permissions when users specify cloudwatch log group through logConfiguration',
+    detailsMd: `
+    Currently, we will automatically add a number of cloudwatch permissions to the task role when no cloudwatch log group is
```
Comment (shikha372) [inline packages/aws-cdk-lib/cx-api/lib/features.ts:1136]:
```suggestion
    Currently, we automatically add a number of cloudwatch permissions to the task role when no cloudwatch log group is
```

**Review by shikha372 — APPROVED** [review]

**Pull request merged**